File: app/src/main/java/com/drdisagree/iconify/config/RPrefs.java
package com.drdisagree.iconify.config;

import static android.content.Context.MODE_PRIVATE;
import static com.drdisagree.iconify.common.Preferences.STR_NULL;
import static com.drdisagree.iconify.common.Resources.SharedXPref;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.SharedPreferences;

import com.drdisagree.iconify.Iconify;

public class RPrefs {

    @SuppressLint("StaticFieldLeak")
    private static final Context prefContext = Iconify.getAppContext().createDeviceProtectedStorageContext();

    static SharedPreferences prefs = prefContext.getSharedPreferences(SharedXPref, MODE_PRIVATE);
    static SharedPreferences.Editor editor = prefs.edit();

    // Save sharedPref config
    public static void putBoolean(String key, boolean val) {
        editor.putBoolean(key, val).commit();
    }

    public static void putInt(String key, int val) {
        editor.putInt(key, val).commit();
    }

    public static void putFloat(String key, float val) {
        editor.putFloat(key, val).commit();
    }

    public static void putString(String key, String val) {
        editor.putString(key, val).commit();
    }

    // Load sharedPref config
    public static boolean getBoolean(String key) {
        return prefs.getBoolean(key, false);
    }

    public static boolean getBoolean(String key, Boolean defValue) {
        return prefs.getBoolean(key, defValue);
    }

    public static int getInt(String key) {
        return prefs.getInt(key, 0);
    }

    public static int getInt(String key, int defValue) {
        return prefs.getInt(key, defValue);
    }

    public static float getFloat(String key) {
        return prefs.getFloat(key, 0);
    }

    public static float getFloat(String key, float defValue) {
        return prefs.getFloat(key, defValue);
    }

    public static String getString(String key) {
        return prefs.getString(key, STR_NULL);
    }

    public static String getString(String key, String defValue) {
        return prefs.getString(key, defValue);
    }

    // Clear specific sharedPref config
    public static void clearPref(String key) {
        editor.remove(key).apply();
    }

    // Clear all sharedPref config
    public static void clearAllPrefs() {
        editor.clear().apply();
    }
}

File: app/src/main/java/com/drdisagree/iconify/config/RPrefs.kt
package com.drdisagree.iconify.config

import android.content.Context.MODE_PRIVATE
import android.content.SharedPreferences
import android.content.SharedPreferences.OnSharedPreferenceChangeListener
import com.drdisagree.iconify.Iconify.Companion.appContext
import com.drdisagree.iconify.common.Resources.SHARED_XPREFERENCES
import com.drdisagree.iconify.ui.preferences.SliderPreference

@Suppress("unused")
object RPrefs : SharedPreferences {

    private val prefs: SharedPreferences by lazy {
        appContext.createDeviceProtectedStorageContext().getSharedPreferences(
            SHARED_XPREFERENCES, MODE_PRIVATE
        )
    }

    private val editor: SharedPreferences.Editor by lazy { prefs.edit() }

    val instance: RPrefs
        get() = this

    val getPrefs: SharedPreferences
        get() = prefs

    // Basic put methods
    fun putBoolean(key: String?, value: Boolean) {
        editor.putBoolean(key, value).apply()
    }

    fun putInt(key: String?, value: Int) {
        editor.putInt(key, value).apply()
    }

    fun putFloat(key: String?, value: Float) {
        editor.putFloat(key, value).apply()
    }

    fun putLong(key: String?, value: Long) {
        editor.putLong(key, value).apply()
    }

    fun putString(key: String?, value: String?) {
        editor.putString(key, value).apply()
    }

    // Basic get methods
    fun getBoolean(key: String?): Boolean {
        return prefs.getBoolean(key, false)
    }

    fun getInt(key: String?): Int {
        return prefs.getInt(key, 0)
    }

    fun getLong(key: String?): Long {
        return prefs.getLong(key, 0)
    }

    fun getFloat(key: String?): Float {
        return prefs.getFloat(key, 0f)
    }

    fun getString(key: String?): String? {
        return prefs.getString(key, null)
    }

    // Custom slider preference methods
    fun getSliderInt(key: String?, defaultVal: Int): Int {
        return SliderPreference.getSingleIntValue(this, key, defaultVal)
    }

    fun getSliderValues(key: String?, defaultValue: Float): List<Float> {
        return SliderPreference.getValues(this, key, defaultValue)
    }

    fun getSliderFloat(key: String?, defaultVal: Float): Float {
        return SliderPreference.getSingleFloatValue(this, key, defaultVal)
    }

    // Clear methods
    fun clearPref(key: String?) {
        editor.remove(key).apply()
    }

    fun clearPrefs(vararg keys: String?) {
        keys.forEach { key ->
            editor.remove(key).apply()
        }
    }

    fun clearAllPrefs() {
        editor.clear().apply()
    }

    // Implementing SharedPreferences interface
    override fun getAll(): Map<String, *> {
        return prefs.all
    }

    override fun getBoolean(key: String?, defValue: Boolean): Boolean {
        return prefs.getBoolean(key, defValue)
    }

    override fun getInt(key: String?, defValue: Int): Int {
        return prefs.getInt(key, defValue)
    }

    override fun getLong(key: String?, defValue: Long): Long {
        return prefs.getLong(key, defValue)
    }

    override fun getFloat(key: String?, defValue: Float): Float {
        return prefs.getFloat(key, defValue)
    }

    override fun getString(key: String?, defValue: String?): String? {
        return prefs.getString(key, defValue)
    }

    override fun getStringSet(key: String, defValues: Set<String>?): Set<String>? {
        return prefs.getStringSet(key, defValues)
    }

    override fun contains(key: String): Boolean {
        return prefs.contains(key)
    }

    override fun edit(): SharedPreferences.Editor {
        return editor
    }

    override fun registerOnSharedPreferenceChangeListener(listener: OnSharedPreferenceChangeListener) {
        prefs.registerOnSharedPreferenceChangeListener(listener)
    }

    override fun unregisterOnSharedPreferenceChangeListener(listener: OnSharedPreferenceChangeListener) {
        prefs.unregisterOnSharedPreferenceChangeListener(listener)
    }
}
